### ISSUE TITLE
reindex and move sentence based solr cluster to amazon

### ISSUE THREAD

**Issue by hroberts** [berkmancenter/mediacloud#352]
I would like to be able to install the story based index in the local mcquery* machines without having to also be running the sentences based index on them.  The easiest way to do that is just to run the sentences solr index on amazon for the next few weeks.  This will also allow us to add a bit more memory to the instances, which should increase memory and reliability. 

Here's the plan:

[] fire up three i3.8xlarge instances (32 core / 244GB / 4 x 1900 NVMe SSD);
[] rsync current index to get most of the data over;
[] shut down imports to production database and do final rsync;
[] do final rsync to get consistent index;
[] point production machines to the remote instance.

Once we are running on aws, we can wipe the local machines and start a new process of importing the story based index on to them.  This will give us some buffer time to optimize the story index with the production data on the local machines.

I think this will cost us around $5k to be running solr in the cloud for a month-ish, but we have the hardware money, and it should make the transition much smoother.

**Comment by hroberts** [comment]
I am running the solr dump on mcnlp now.

**Comment by hroberts** [comment]
Updated original comment to just rsync current index files rather than trying to reindex from postgres.

The export script run on mcnlp failed because it ran out of memory.  The new story based code does not have the same problem with requiring obscene amounts of memory to export (both because there are fewer stories than sentences and because the code no longer does the client side joins).  Rather than fight with the legacy code to make it work sanely, I am going to just use the existing solr indexes and sacrifice having the docvalues fields.  If we keep getting the crashes even with the increased memory in the solr instances I can try reimplementing the date and stories counts in ways that don't use grouping or faceting.

**Comment by hroberts** [comment]
This took a lot more fiddling than I thought it would, but the sentence index cluster is now running on amazon.  The performance seems much better in casual testing for some reason (many queries are 5-10x faster).  I did not reindex but instead just rsync'd the existing index to the amazon machines.  The amazon machine have 244G instead of the 192G of our local machines, but I can't imagine that little bit of extra memory makes that much of a difference.  Will have to investigate further.

Most important thing to note about the current setup is that we are running solr off of the ssd instance drives because they are reportedly much faster than the ebs ssd drives.  That means that if we shut down any of the instances rather than merely rebooting them, all of the solr data will disappear and we will have to rsync over from backup, which takes a couple of hours.  So don't shutdown these instances.

Here are some notes on the amazon solr setup:
* we are running on 3 is.8xlarge hosts;
* the machines each have the default ubuntu user and a mediacloud user;
* to connect, use the bkmn.pem key to the ubuntu user;
* I also created a separate key in  mccore1:.ssh/aws_rsa that I've been using to do rsyncs and such;
* the machines have the public names awsquery1.haldir.org, awsquery2,haldir.org, and awsquery4.haldir.org and were rsync'd from the corresponding mcquery* machines;
* each machines has a local host name of 'awsquery*', and the private aws vpc ip for each of those names is in /etc/hosts for each machine;
* the mediacloud home dir is on the root ebs store but the mediacloud repo and all solr data is on the ssd instance drives;
* the ssd instance drives are setup as md raid10 under /dev/md0, which is mounted as /space;
* /space/mediacloud-mit is a rsync'd copy of mcquery*:/home/mediacloud/mediacloud;
* python didn't like it when I tried to run mediacloud directly from that rsync'd dir, so I installed a fresh mediacloud repo into /space/mediacloud;
* /home/mediacloud/mediacloud is a link to /space/mediacloud;
* /space/mediacloud/data is a link to /space/mediacloud-mit/data;
* each machine is runhing a ufw firewall to only allow access to the solr ports from other machines on the vpc private ip block;
* /etc/security/limits.d/50-mediacloud.conf has to be edited to allow unlimited virtual memory;
* before starting up all of the local shards, I used the following bash script to prune all existing mcquery* shard replicants;
* I found using parallel rsyncs from mit to amazon to be much faster than a single rsync.

```
for c in 1 2; do
    for s in `seq 1 24`; do 
        for n in `seq 1 100`; do 
            curl "http://localhost:7983/solr/admin/collections?action=DELETEREPLICA&collection=collection$c&shard=shard$s&replica=core_node$n"; 
        done
    done; 
done
```
*  I setup an ssh tunnel running mcquery3-private:7983 to the amazon machines and edited mediawords.yml on mccore1 and mcquery3 to use 'http://mcquery3-private:7983/solr/' as the solr_url.

I am going to let this run in production for a couple hours before turning solr imports back on on mccore1.

**Comment by pypt** [comment]
Wow, that was a lot of hassle!
